### PR TITLE
Reflectometry GUI restructuring 9: set defaults

### LINKED ISSUE #23046: ISIS Reflectometry Interface - Make Default Value Loading Consistent across GUI and Algorithms
*This issue refers to work which should be based on changes currently only present in the branch `isis_reflectometry_alternative_table`.*

There is some ambiguity over the expected behaviour in the event of empty/default values. Some
default's exist in different levels of the GUI, some exist in the algorithms. The behaviour should be unionised among all of these as much as possible. We should then be able to retrieve default values via an instance of the model (created from an algorithm instance) and injected into the presenter on construction.

The current suggested approach with changes to the GUI is somewhat reliant upon the interface being able to obtain a set of default values for all fields at the time when the interface is launched, this is because, in switching numeric fields to `QSpinBox`s it has become difficult for numeric inputs to be empty. If this constraint proves problematic we could move back to using `QLineEdits`, this shifts more custom validation back into the presenters though.

The current suggestion is to try to load all settings from the parameter file and show
an error, preventing usage if this is not possible for any non-optional settings. This has not yet
been cleared with all instrument scientists.

It would also be nice to hide inputs when they are not going to be used - currently they are greyed out but may still show values that don't make sense for the current selection, e.g. change SummationType to SumInQ, check IncludePartialBins and change SummationType back. IncludePartialBins remains checked but it would be better if it were hidden.

### PR DESCRIPTION
**Pull request by gemmaguest**
~~***PR #25605 should be merged first***~~

**Description of work.**
Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR adds functionality to populate settings on the GUI with defaults taken from the reduction algorithm or parameters file.

- The changes to files in `API` and `Algorithms` are due to moving `BoostOptionalToAlgorithmProperty` to  `API` to make it available from the GUI.
- The changes to `ReflectometryHelper` to use the `AlgorithmManager` were necessary to use it from the GUI tests.
- Release notes have been added for all changes on the feature branch so far, so they include a lot more than just the content of this PR. The relevant part to this PR is that the defaults are now set automatically on the Experiment and Instrument tabs. I added the rest because this change makes more sense in the overall context.

**To test:**

- Set your default facility to `ISIS` and default instrument to `INTER`
- Open the `ISIS Reflectometry` interface
- Check that the populated values in the `Experiment` and `Instrument` tabs match the values in the `INTER_parameters.xml` file:

image

- On the `Runs` tab, change the instrument to `OFFSPEC` and check that the values have been updated to match the `OFFSPEC_Parameters.xml` file.

image

- Note that at the moment if you toggle to the `Batch 1` tab on the left or add new batches, they will have the original instrument (INTER). We may change this in future but for now just check that the defaults still match the instrument for the batch.
- Change some values and then click `Restore Defaults`. Check that values from the parameters file are re-instated and any error highlighting is cleared. 
- Edit `INTER_Parameters.xml` in your local instrument directory (e.g. `~/.mantid/instrument` on Linux or `%APPDATA%/mantidproject/mantid/instrument` on Windows). Edit some values to make them invalid e.g. set `LambdaMax` to be less than `LambdaMin`. Restart mantid and open the interface. Check that an error is reported in the log and that the relevant inputs on the interface are unset.
  - Note that entering invalid types in the parameters file e.g. a string value when a numeric value is expected, is currently treated as unset rather than reporting an error; this is perhaps not ideal but is due to the way that `BoostOptionalToAlgorithmProperty` works and changing that is beyond the scope of this PR.

Fixes #23046 and Refs #23051

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/test/ISISReflectometry/Instrument/InstrumentOptionDefaultsTest.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
```
Comment (alicerussell1) [inline qt/scientific_interfaces/test/ISISReflectometry/Instrument/InstrumentOptionDefaultsTest.h]:
```suggestion
// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
```

File: qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentOptionDefaults.cpp [hunk 2]
```diff
@@ -5,56 +5,100 @@
...
+  auto transmissionRunRange = RangeInLambda(
+      defaults.getDoubleOrZero("StartOverlap", "TransRunStartOverlap"),
+      defaults.getDoubleOrZero("EndOverlap", "TransRunEndOverlap"));
+
+  if (!transmissionRunRange.isValid(false))
+    throw std::invalid_argument("Transmission run overlap range is invalid");
+
+  // We currently don't specify stitch parameters in the parameters file
+  // although we
+  auto stitchParameters = std::map<std::string, std::string>();
```
Comment (alicerussell1) [inline qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentOptionDefaults.cpp]: This comment is unfinished, is this going to be specified in the future?

**Review by alicerussell1 — CHANGES_REQUESTED** [review]
The code looks mostly good however I have noticed one thing. When on the experiment settings tab if you change the summation type to `SumInQ` and change another property e.g. `PolarisationCorrections` and then press restore defaults it will only change the summation type. If you press restore again it will then change the remaining properties to their defaults.

File: qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentOptionDefaults.cpp (hunk 2 above)
Comment (gemmaguest) [reply qt/scientific_interfaces/ISISReflectometry/GUI/Experiment/ExperimentOptionDefaults.cpp]: I can't remember what I was going to write here so I've removed the unfinished bit

**Comment by gemmaguest** [comment]
I think I've addressed all the points so this should be ready for another look, assuming tests pass ok. Good spot on the summation type change bug - this was because it has its own slot which was not being disconnected when we update the view, so it was notifying back to the presenter and overwriting the model.

**Review by alicerussell1 — APPROVED** [review]
This is now looking good 👍

**Comment by gemmaguest** [comment]
The force push was to rebase onto the destination branch to fix conflicts

**Comment by martyngigg** [comment]
I'll merge it once the builds finish.

**Pull request merged**